### PR TITLE
bpo-31368: Enhance os.preadv() documentation

### PR DESCRIPTION
**Pull request by vstinner**
https://bugs.python.org/issue31368

### CODE REVIEW

**Comment by vstinner** [comment]
@pablogsal: would you mind to review my PR?

cc @YoSTEALTH @njsmith @csabella

File: Doc/library/os.rst [hunk 1]
```diff
@@ -1104,39 +1104,48 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
...
-   *buffers* must be a sequence of :term:`bytes-like objects <bytes-like object>`.
-   Buffers are processed in array order. Entire contents of first buffer is written
-   before proceeding to second, and so on. The operating system may set a limit
-   (sysconf() value SC_IOV_MAX) on the number of buffers that can be used.
-   :func:`~os.pwritev` writes the contents of each object to the file descriptor
-   and returns the total number of bytes written.
+   Write the *buffers* contents to file descriptor *fd* at offset *offset*.
+   *buffers* must be a sequence of :term:`bytes-like objects <bytes-like
+   object>`. Buffers are processed in array order. Entire contents of first
+   buffer is written before proceeding to second, and so on.
```
Comment (csabella) [inline Doc/library/os.rst]: 'Entire contents of ***the*** first buffer'
'before proceeding to **the** second'

Comment (vstinner) [reply Doc/library/os.rst]: done

File: Doc/library/os.rst [hunk 2]
```diff
@@ -1328,9 +1350,10 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
 
    Write the contents of *buffers* to file descriptor *fd*. *buffers* must be a
    sequence of :term:`bytes-like objects <bytes-like object>`. Buffers are
-   processed in array order. Entire contents of first buffer is written before
-   proceeding to second, and so on. The operating system may set a limit
-   (sysconf() value SC_IOV_MAX) on the number of buffers that can be used.
+   processed in array order. Entire contents of the first buffer is written
+   before proceeding to the second, and so on. The operating system may set a
+   limit (sysconf() value SC_IOV_MAX) on the number of buffers that can be
```
Comment (pablogsal) [inline Doc/library/os.rst]: It will make sense to change 
```
 sysconf() -> :func:`sysconf` 
```
and
```
SC_IOV_MAX ->  ``'SC_IOV_MAX'``
```
?

**Review by pablogsal — CHANGES_REQUESTED** [review]

File: Doc/library/os.rst [hunk 3]
```diff
@@ -1092,6 +1092,64 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
...
+
+   Combine the functionality of :func:`os.readv` and :func:`os.pread`.
+
+   The flags argument contains a bitwise OR of zero or more of the following
+   flags:
+
+   - :data:`RWF_HIPRI`
+   - :data:`RWF_NOWAIT`
+
+   Using non-zero flags requires Linux 4.6 or newer.
```
Comment (YoSTEALTH) [inline Doc/library/os.rst]: could be change this to: "Using non-zero flags requires preadv2 support on Linux 4.7 or newer."

File: Doc/library/os.rst [hunk 4]
```diff
@@ -1104,39 +1162,48 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
...
 
    The *flags* argument contains a bitwise OR of zero or more of the following
    flags:
 
-   - RWF_DSYNC
-   - RWF_SYNC
+   - :data:`RWF_DSYNC`
+   - :data:`RWF_SYNC`
 
    Using non-zero flags requires Linux 4.7 or newer.
```
Comment (YoSTEALTH) [inline Doc/library/os.rst]: could be also change this to: "Using non-zero flags requires pwritev2 support on Linux 4.7 or newer."

File: Doc/library/os.rst [hunk 5]
```diff
@@ -1092,6 +1092,64 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
...
+
+   - :data:`RWF_HIPRI`
+   - :data:`RWF_NOWAIT`
+
+   Using non-zero flags requires Linux 4.6 or newer.
+
+   The operating system may set a limit (:func:`sysconf` value
+   ``'SC_IOV_MAX'``) on the number of buffers that can be used.
+
+   Availability: Linux 2.6.30 abd newer, FreeBSD 6.0 and newer,
```
Comment (YoSTEALTH) [inline Doc/library/os.rst]: should be "Availability: Linux 2.6.30, FreeBSD 6.0 and newer."

File: Doc/library/os.rst (hunk 2 above)
Comment (vstinner) [reply Doc/library/os.rst]: Oh wait, I didn't want to modify this documentation... But then I read the doc, and oh, all these documentations can be enhanced...

**Comment by vstinner** [comment]
Sorry for reviewers, but I basically rewrote my whole PR to cleanup the documentation of all read and write functions:
* read, pread, readv, preadv
* write, pwrite, writev, pwritev

The documentation should now be more consistent.

**Review by YoSTEALTH — CHANGES_REQUESTED** [review]

File: Doc/library/os.rst (hunk 3 above)
Comment (vstinner) [reply Doc/library/os.rst]: I'm trying to avoid mentioning preadv2(). Do you think that you must mention it?

File: Doc/library/os.rst (hunk 5 above)
Comment (vstinner) [reply Doc/library/os.rst]: "and newer" is commonly used in os.rst documentation (i fixed the typo)

File: Doc/library/os.rst (hunk 3 above)
Comment (YoSTEALTH) [reply Doc/library/os.rst]: All the flags actually belong to preadv2() function though, maybe a tiny mention wont hurt?
Its your call.

Comment (njsmith) [reply Doc/library/os.rst]: You should absolutely mention `preadv2` and `pwritev2`. At the point someone is using these features in the `os` module they're also reading man pages and lwn articles, and you want to help them make the connection. If someone wonders whether python has a wrapper for `preadv2` then searching for that in the docs should find `os.preadv`.

File: Doc/library/os.rst [hunk 6]
```diff
@@ -1082,20 +1082,81 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
...
+   - :data:`RWF_HIPRI`
+   - :data:`RWF_NOWAIT`
+
+   Return the total number of bytes actually read which can be less than the
+   total capacity of all the objects.
+
+   The operating system may set a limit (:func:`sysconf` value
+   ``'SC_IOV_MAX'``) on the number of buffers that can be used.
+
+   Combine the functionality of :func:`os.readv` and :func:`os.pread`.
```
Comment (vadmium) [inline Doc/library/os.rst:1117]: This reads as an instruction, but it is no longer clear it is for the implementation. Maybe drop the imperative and write “This function combines the functionality . . .”. Also applies to _pwritev_ below.

File: Doc/library/os.rst [hunk 7]
```diff
@@ -1082,20 +1082,81 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
...
+   move on to the next buffer in the sequence to hold the rest of the data.
+
+   The flags argument contains a bitwise OR of zero or more of the following
+   flags:
+
+   - :data:`RWF_HIPRI`
+   - :data:`RWF_NOWAIT`
+
+   Return the total number of bytes actually read which can be less than the
+   total capacity of all the objects.
```
Comment (vadmium) [inline Doc/library/os.rst:1112]: Needs a comma: “. . . number of bytes read, which can be less . . .”. Applies to _readv_ below too.

File: Doc/library/os.rst [hunk 8]
```diff
@@ -1082,20 +1082,81 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
...
+
+.. data:: RWF_NOWAIT
+
+   Do not wait for data which is not immediately available. If this flag is
+   specified, the system call will return instantly if it would have to read
+   data from the backing storage or wait for a lock.
+
+   If some data was successfully read, it will return the number of bytes read.
+   If no bytes were read, it will return ``-1`` and set errno to
+   :data:`errno.EAGAIN`.
```
Comment (vadmium) [inline Doc/library/os.rst:1133]: I’m confused. Normally you access _errno_ as an attribute of an exception. Do the Python wrapper(s) return −1 or raise OSError/BlockingIOError?

File: Doc/library/os.rst [hunk 9]
```diff
@@ -1104,48 +1165,57 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
...
 
-   Combines the functionality of :func:`os.writev` and :func:`os.pwrite`. It
-   writes the contents of *buffers* to file descriptor *fd* at offset *offset*.
-   *buffers* must be a sequence of :term:`bytes-like objects <bytes-like object>`.
-   Buffers are processed in array order. Entire contents of first buffer is written
-   before proceeding to second, and so on. The operating system may set a limit
-   (sysconf() value SC_IOV_MAX) on the number of buffers that can be used.
-   :func:`~os.pwritev` writes the contents of each object to the file descriptor
-   and returns the total number of bytes written.
+   Write the *buffers* contents to file descriptor *fd* at a offset *offset*,
```
Comment (vadmium) [inline Doc/library/os.rst:1168]: _a_ → _an_ or _the_, or even better “at offset _offset_” would work I think.

**Comment by vstinner** [comment]
I abandon my change. Feel free to reuse my change to create a new PR if you want to enhance the documentation.

**Pull request closed without merging**